homeassistant/components/eq3btsmart/climate.py

```diff
@@ -0,0 +1,422 @@
+            self._on_device_updated()
+
+        self.schedule_update_ha_state()
+
+    @callback
+    def _on_status_updated(self):
+        """Handle updated status from the thermostat."""
+
+        self._target_temperature = self._thermostat.status.target_temperature.value
+        self._attr_hvac_mode = EQ_TO_HA_HVAC[self._thermostat.status.operation_mode]
+
+    @callback
+    def _on_device_updated(self):
+        """Handle updated device data from the thermostat."""
+
+        self._firmware_version = str(self._thermostat.device_data.firmware_version)
+
+        device_registry = async_get(self.hass)
+        device = device_registry.async_get_device(
+            identifiers={(DOMAIN, self._eq3_config.mac_address)},
+        )
+
+        if device:
+            device_registry.async_update_device(
+                device_id=device.id,
+                sw_version=self._firmware_version,
+            )
+
+    @callback
+    def _on_connection_changed(self, is_connected: bool = True) -> None:
+        """Handle connection changed."""
+
+        self._is_available = is_connected
+
+        if is_connected and not self._was_connected:
+            _LOGGER.info("[%s] Connected", self._eq3_config.name)
+            self._was_connected = True
+            self.hass.add_job(self._update_device())
+
+        if not is_connected and self._was_connected:
+            _LOGGER.warning("[%s] Disconnected", self._eq3_config.name)
+            self._was_connected = False
+
+        self.schedule_update_ha_state()
+
+    @callback
+    async def _update_device(self) -> None:
+        try:
+            await self._thermostat.async_get_id()
+        except Eq3Exception:
+            _LOGGER.error(
+                "[%s] Error fetching device information", self._eq3_config.name
+            )
+            return
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+
+        return self._is_available
+
+    @property
+    def hvac_action(self) -> HVACAction | None:
+        """Return the current running hvac operation."""
+
+        if self._thermostat.status is None:
+            return None
+
+        if self._thermostat.status.operation_mode == OperationMode.OFF:
+            return HVACAction.OFF
+
+        if self._thermostat.status.valve == 0:
+            return HVACAction.IDLE
+
+        return HVACAction.HEATING
+
+    @property
+    def current_temperature(self) -> float | None:
+        """Return the current temperature."""
+
+        match self._eq3_config.current_temp_selector:
+            case CurrentTemperatureSelector.NOTHING:
+                return None
+            case CurrentTemperatureSelector.VALVE:
+                if self._thermostat.status is None:
+                    return None
+
+                return (
+                    (1 - self._thermostat.status.valve / 100) * 2
+                    + self._thermostat.status.target_temperature.value
+                    - 2
```

Comment:
I think we're not supposed to do calculations in our entites, but only in the lib

Reply:
Done, calculation is now implemented in the library.